### PR TITLE
more element tests, fixed found bugs

### PR DESCRIPTION
**Pull request by mat-hek**

### CODE REVIEW

File: test/membrane/core/element_test.exs [hunk 1]
```diff
@@ -11,11 +11,225 @@ defmodule Membrane.Core.ElementTest do
...
+               Element.handle_call(msg, nil, get_state())
+    end)
+
+    assert {:reply, {:error, {:cannot_handle_message, {:invalid_message, :abc, _}, _}}, _state} =
+             Element.handle_call(:abc, nil, get_state())
+  end
+
+  test "other message" do
+    state = get_state()
+    assert {:noreply, state} == Element.handle_info(:abc, state)
```
Comment (bblaszkow06) [inline test/membrane/core/element_test.exs]: Shouldn't that verify that `handle_other` was actually called?

**Comment by bblaszkow06** [comment]
I guess this should be merged to master, not `bins_merge`, right?

File: test/membrane/core/element_test.exs [hunk 2]
```diff
@@ -11,11 +11,225 @@ defmodule Membrane.Core.ElementTest do
...
+  end
+
+  test "demand unit" do
+    assert {:noreply, state} =
+             Element.handle_info(Message.new(:demand_unit, [:bytes, :output]), linked_state())
+
+    assert state.pads.data.output.other_demand_unit == :bytes
+  end
+
+  test "demand or buffer or caps or event when not playing" do
```
Comment (kanes115) [inline test/membrane/core/element_test.exs]: Again, more of an opinion: ExUnit gives the opportunity to define test case names as strings which is powerful. I don't like trying to shorten this and not creating sentences (add a predicate!). I can see this notion throughout the code base :( Look how horrible [this gets in erlang](https://github.com/esl/MongooseIM/blob/master/big_tests/tests/mam_SUITE.erl#L78) and still this is a sentence :D

File: test/membrane/core/element_test.exs [hunk 3]
```diff
@@ -11,11 +11,225 @@ defmodule Membrane.Core.ElementTest do
...
+      Message.new(:event, %Membrane.Testing.Event{}, for_pad: :input),
+      Message.new(:event, %Membrane.Testing.Event{}, for_pad: :output)
+    ]
+    |> Enum.each(fn msg ->
+      assert {:noreply, state} = Element.handle_info(msg, initial_state)
+      assert {:ok, state} == Element.PlaybackBuffer.store(msg, initial_state)
+    end)
+  end
+
+  test "demand" do
```
Comment (kanes115) [inline test/membrane/core/element_test.exs]: This test does not test a lot. This only tests the situation in which callback is not implemented. This is fine of course, but I feel this should be reflected in test name as now it sounds like it tests the general happy path.

Comment (kanes115) [reply test/membrane/core/element_test.exs]: As discussed in private.
I think this callback call should be mocked somehow (preferably there should be a short callback module injected from test).

File: test/membrane/core/element_test.exs [hunk 4]
```diff
@@ -11,11 +11,225 @@ defmodule Membrane.Core.ElementTest do
...
+  end
+
+  test "handle unlink" do
+    assert {:noreply, state} =
+             Element.handle_info(Message.new(:handle_unlink, :input), linked_state())
+
+    refute Map.has_key?(state.pads.data, :input)
+  end
+
+  test "timer tick" do
```
Comment (kanes115) [inline test/membrane/core/element_test.exs]: Those kind of tests will pass even if Element implements `handle_info` fallback and not do anything. This imo makes those tests hardly useful (I believes we should assert something more).

File: test/membrane/core/element_test.exs [hunk 5]
```diff
@@ -11,11 +11,225 @@ defmodule Membrane.Core.ElementTest do
...
+    ]
+    |> Enum.each(fn msg ->
+      assert {:stop, {:error, {:cannot_handle_message, {:invalid_message, ^msg, _}, _}}, _state} =
+               Element.handle_info(msg, get_state())
+
+      assert {:reply, {:error, {:cannot_handle_message, {:invalid_message, ^msg, _}, _}}, _state} =
+               Element.handle_call(msg, nil, get_state())
+    end)
+
+    assert {:reply, {:error, {:cannot_handle_message, {:invalid_message, :abc, _}, _}}, _state} =
```
Comment (kanes115) [inline test/membrane/core/element_test.exs]: Why there's no `handle_cast`?

File: lib/membrane/core/element/message_dispatcher.ex [hunk 6]
```diff
@@ -118,7 +118,7 @@ defmodule Membrane.Core.Element.MessageDispatcher do
     {:ok, new_state}
   end
 
-  defp do_handle_message(Message.new(_, _) = message, mode, state) do
+  defp do_handle_message(Message.new(_, _, _) = message, mode, state) do
```
[the remark this answers is not in the document]
Comment (mat-hek) [reply lib/membrane/core/element/message_dispatcher.ex]: ok, will edit the commit message

File: test/membrane/core/element_test.exs (hunk 5 above)
Comment (mat-hek) [reply test/membrane/core/element_test.exs]: because we don't use it at all :)

File: test/membrane/core/element_test.exs (hunk 1 above)
Comment (mat-hek) [reply test/membrane/core/element_test.exs]: These tests are for the element core, i.e. GenServer behind the element. I think checks for callbacks being called should be in proper controllers or integration tests

**Pull request merged**